### PR TITLE
Swift-28 Implement Sessions

### PR DESCRIPTION
**Pull request by patrickfreed**
This PR adds the implementation of client sessions as per the [design](https://docs.google.com/document/d/1n5N8f1Zgzd5Iow5gxiRSoj6FZ8-Gus8I5CUTrW1Gh3s/edit#heading=h.ky3pv4dt04e).

Deviations from the design:
- `session.sessionId` refactored to `session.id`, the "session" prefix seemed redundant
- `session.endSession` -> `session.end` ditto redundancy
- The test plan prescribed performing spec test 9 (ensure all sessions are cleaned up after each test), but without access to the underlying pool this is difficult
- The test plan states that many of the tests should run every function that takes in a session or against every topology with every read preference. This is overkill for what we're implementing. Assuming libmongoc is performing those tests at their level, all we need to do is ensure that every method that accepts a session properly passes it to libmongoc, which we verify in the `testSessionArguments` test. For the other tests, we just use a single method that takes in a session as a representative.
- I removed the `ClientSessionPointer` typealias, since libmongoc doesn't actually export the complete `mongoc_client_session_t` type. Regardless of compiler version, pointers to it will be imported to swift as `OpaquePointer`s. Since those pointers will always be mutable, I only defined `MutableClientSessionPointer`.

### CODE REVIEW

File: Sources/MongoSwift/BSON/Document.swift [hunk 1]
```diff
@@ -54,7 +54,7 @@ public struct Document {
     internal var storage: DocumentStorage
 
     /// Returns the number of (key, value) pairs stored at the top level of this `Document`.
-    public fileprivate(set) var count: Int
+    public internal(set) var count: Int
```
Comment (patrickfreed) [inline Sources/MongoSwift/BSON/Document.swift]: This change (and the one below) are necessary to use this stuff in `append`, in the future i imagine they'll go back to being private or fileprivate when we work on the memory safety epic.

File: Sources/MongoSwift/MongoClient.swift [hunk 2]
```diff
@@ -270,14 +278,17 @@ public class MongoClient {
...
      *   - `UserError.invalidArgumentError` if the options passed are an invalid combination.
+     *   - `UserError.logicError` if the provided session is inactive.
      *   - `EncodingError` if an error is encountered while encoding the options to BSON.
      */
-    public func listDatabases(options: ListDatabasesOptions? = nil) throws -> MongoCursor<Document> {
-        let opts = try self.encoder.encode(options)
-        guard let cursor = mongoc_client_find_databases_with_opts(self._client, opts?.data) else {
+    public func listDatabases(options: ListDatabasesOptions? = nil,
+                              session: ClientSession? = nil) throws -> MongoCursor<Document> {
+        var opts = try self.encoder.encode(options) ?? Document()
```
Comment (patrickfreed) [inline Sources/MongoSwift/MongoClient.swift]: This (and all the others like it) construct a new `Document` every time, regardless of whether any options or a session is passed in. I think `bson_t` does some nifty things where this is a stack allocation if the doc is small enough, so I don't think this is a big performance issue.

File: Sources/MongoSwift/MongoCollection+FindAndModify.swift [hunk 3]
```diff
@@ -383,6 +391,18 @@ private class FindAndModifyOptions {
...
     }
 
+    fileprivate func setSession(_ session: ClientSession?) throws {
+        guard let session = session else {
+            return
+        }
+        var doc = Document()
+        try session.append(to: &doc)
+
+        guard mongoc_find_and_modify_opts_append(self._options, doc.data) else {
```
Comment (patrickfreed) [inline Sources/MongoSwift/MongoCollection+FindAndModify.swift]: this is a very silly api provided by libmongoc

File: Sources/MongoSwift/MongoCollection+Indexes.swift [hunk 4]
```diff
@@ -163,13 +163,17 @@ extension MongoCollection {
...
      *   - `UserError.invalidArgumentError` if the options passed in form an invalid combination.
+     *   - `UserError.logicError` if the provided session is inactive.
      *   - `EncodingError` if an error occurs while encoding the index specification or options.
      */
     @discardableResult
     public func createIndex(_ keys: Document,
                             options: IndexOptions? = nil,
-                            commandOptions: CreateIndexOptions? = nil) throws -> String {
-        return try createIndexes([IndexModel(keys: keys, options: options)], options: commandOptions)[0]
+                            commandOptions: CreateIndexOptions? = nil,
```
Comment (patrickfreed) [inline Sources/MongoSwift/MongoCollection+Indexes.swift]: I noticed this was called `commandOptions` here, but just `options` in the other index methods. Was that intentional?

File: Sources/MongoSwift/Operations/DropIndexesOperation.swift [hunk 5]
```diff
@@ -14,24 +14,29 @@ public struct DropIndexOptions: Encodable {
 /// An operation corresponding to a "dropIndexes" command.
 internal struct DropIndexesOperation<T: Codable>: Operation {
     private let collection: MongoCollection<T>
+    private let session: ClientSession?
```
Comment (patrickfreed) [inline Sources/MongoSwift/Operations/DropIndexesOperation.swift]: this is a bit out of scope for this work, but I imagine the operations, when we do client pooling, will try to get a collection via the session first, then from the stored mongo collection second if there is no stored session.

File: Sources/MongoSwift/MongoClient.swift (hunk 2 above)
Comment (mbroadst) [reply Sources/MongoSwift/MongoClient.swift]: I know we kind of duked this out on the design already, but something makes me uncomfortable about this approach. Could we instead make all of the options conform to some protocol that provided a `withSession(ClientSession?)`, so you would only add it in the event that the options did exist (and therefore only allocate once)? Something like:
```
guard let cursor = mongoc_client_find_databases_with_opts(self._client, opts?.withSession(session).data) else {
```

Originally I was thinking of `SessionProvider` storing the session with the options, but this would be like an "appending copy" of the options.

File: Sources/MongoSwift/MongoCollection+FindAndModify.swift (hunk 3 above)
Comment (mbroadst) [reply Sources/MongoSwift/MongoCollection+FindAndModify.swift]: 😞

File: Sources/MongoSwift/MongoCollection+Indexes.swift [hunk 6]
```diff
@@ -306,13 +327,20 @@ extension MongoCollection {
...
      * - Returns: A `MongoCursor` over the index names.
      *
-     * - Throws: A `userError.invalidArgumentError` if the options passed are an invalid combination.
+     * - Throws: `UserError.logicError` if the provided session is inactive.
      */
-    public func listIndexes() throws -> MongoCursor<Document> {
-        guard let cursor = mongoc_collection_find_indexes_with_opts(self._collection, nil) else {
+    public func listIndexes(session: ClientSession? = nil) throws -> MongoCursor<Document> {
+        var opts: Document?
+        if let session = session {
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoCollection+Indexes.swift]: Is this the only place we have a method without options? Perhaps we should encapsulate this in some sort of utility method

File: Sources/MongoSwift/MongoCursor.swift [hunk 7]
```diff
@@ -16,13 +18,19 @@ public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
      *   - `UserError.invalidArgumentError` if the options passed to the command that generated this cursor formed an
      *     invalid combination.
      */
-    internal init(fromCursor cursor: OpaquePointer,
-                  withClient client: MongoClient,
-                  withDecoder decoder: BSONDecoder) throws {
+    internal init(from cursor: OpaquePointer,
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoCursor.swift]: while we're here and changing stuff, can you add a typealias for the cursor pointer?

File: Sources/MongoSwift/MongoError.swift [hunk 8]
```diff
@@ -148,19 +152,26 @@ internal func parseMongocError(_ error: bson_error_t, errorLabels: [String]? = n
     switch (domain, code) {
     case (MONGOC_ERROR_CLIENT, MONGOC_ERROR_CLIENT_AUTHENTICATE):
         return RuntimeError.authenticationError(message: message)
+    case (MONGOC_ERROR_CLIENT, MONGOC_ERROR_CLIENT_SESSION_FAILURE):
+        // If user attempts to start a session against a server that doesn't support it, we throw a compat error.
+        if message.lowercased().contains("support sessions") {
+            return RuntimeError.compatibilityError(message: "Deployment does not support sessions")
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoError.swift]: nice 👌

File: Sources/MongoSwift/MongoError.swift [hunk 9]
```diff
@@ -148,19 +152,26 @@ internal func parseMongocError(_ error: bson_error_t, errorLabels: [String]? = n
     switch (domain, code) {
     case (MONGOC_ERROR_CLIENT, MONGOC_ERROR_CLIENT_AUTHENTICATE):
         return RuntimeError.authenticationError(message: message)
+    case (MONGOC_ERROR_CLIENT, MONGOC_ERROR_CLIENT_SESSION_FAILURE):
+        // If user attempts to start a session against a server that doesn't support it, we throw a compat error.
+        if message.lowercased().contains("support sessions") {
```
Comment (mbroadst) [inline Sources/MongoSwift/MongoError.swift]: any way to check by error code rather than string match?

File: Sources/MongoSwift/Operations/DropIndexesOperation.swift (hunk 5 above)
Comment (mbroadst) [reply Sources/MongoSwift/Operations/DropIndexesOperation.swift]: I think we should only ever need a namespace, no need to eventually store a collection or a database

File: Sources/MongoSwift/MongoCollection+Indexes.swift (hunk 4 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+Indexes.swift]: Oh whoops, I didn't see the options parameter in there as well. Makes sense.

File: Tests/MongoSwiftTests/ClientSessionTests.swift [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+import Nimble
+import XCTest
+
+final class ClientSessionTests: MongoSwiftTestCase {
+    override func tearDown() {
+        do {
+            let client = try MongoClient(MongoSwiftTestCase.connStr)
+            try client.db(type(of: self).testDatabase).drop()
+        } catch {
+            fail("encountered error when tearing down: \(error)")
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/ClientSessionTests.swift]: Apparently. I copied it from the bulk write tests: https://github.com/mongodb/mongo-swift-driver/blob/master/Tests/MongoSwiftTests/MongoCollection%2BBulkWriteTests.swift#L38

File: Tests/MongoSwiftTests/ClientSessionTests.swift [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+                expect(nestedSession.id).to(bsonEqual(idB))
+            }
+        }
+    }
+
+    struct SessionsArgTest {
+        let name: String
+        let body: (ClientSession?) throws -> Void
+
+        func execute(session: ClientSession) throws {
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Tests/MongoSwiftTests/ClientSessionTests.swift]: Done

File: Sources/MongoSwift/ClientSession.swift [hunk 12]
```diff
@@ -1,35 +1,115 @@
...
+            return
+        }
+        self._opts = mongoc_session_opts_new()
     }
 
-    /// Cleans up internal state.
     deinit {
+        guard let opts = self._opts else {
+            return
+        }
```
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/ClientSession.swift]: We probably should have, but as part of my changes in response to Matt's comments removed this class.

File: Sources/MongoSwift/ClientSession.swift [hunk 13]
```diff
@@ -1,35 +1,106 @@
...
+    var opts = mongoc_session_opts_new()! // always returns a value
+    defer { mongoc_session_opts_destroy(opts) }
+    return try body(opts)
 }
 
-/// :nodoc: A session for ordering sequential operations.
-public class ClientSession: Encodable {
-    /// Initializes a new client session.
-    internal init() {
+/// A MongoDB client session.
```
Comment (mbroadst) [inline Sources/MongoSwift/ClientSession.swift]: Since this is a public type I think we should expand the documentation, or at least link to relevant documentation like [this](https://docs.mongodb.com/manual/reference/server-sessions/).

File: Sources/MongoSwift/ClientSession.swift [hunk 14]
```diff
@@ -1,35 +1,106 @@
...
 /// Options to use when creating a ClientSession.
-internal struct SessionOptions: Encodable {
-    /// Specifies whether read operations should be causally ordered within the session.
-    private let causalConsistency: Bool?
+public struct ClientSessionOptions {}
+
+/// Private helper for providing a `mongoc_session_opt_t` that is only valid within the body of the provided
+/// closure.
+private func withSessionOpts<T>(wrapping options: ClientSessionOptions?,
+                                _ body: (OpaquePointer) throws -> T) rethrows -> T {
```
[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/ClientSession.swift]: I'd actually argued against these before. see [here](https://github.com/mongodb/swift-mongo-mobile/pull/14#discussion_r268257188).

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/ClientSession.swift]: as an addition to that comment, I would say that it was useful to typealias in the cases where a pointer's type is different across Swift versions, for the bson_ts and iterators. I'm just a bit skeptical about ending up with like 30 typealiases that all are ultimately just `OpaquePointer`s.

[the remark this answers is not in the document]
Comment (mbroadst) [reply Sources/MongoSwift/ClientSession.swift]: I think my primary motivation is readability here. We don't actually _lose_ anything from type aliasing, and I think we gain an ability to keep our Swift code in line with what's going on in mongoc. If the goal of keeping it an `OpaquePointer` is to warn the developer that they need to go open the mongoc docs and triple check everything when dealing with it, then we could name it `UnsafeMongocSessionOptsPointer` or something - even that gives me more context than `OpaquePointer`. Are there other benefits to keeping it `OpaquePointer`?

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/ClientSession.swift]: I agree that it increases readability to know what the pointer is supposedly pointing to. I just feel weird about the fact that any unimported libmongoc type can be implicitly cast to a `SessionOptsPointer`.
I think it's easy to be like "oh, this function takes in a real type, and it compiles when I pass in this variable, so my variable must be the right type" even though you passed in a pointer to a bson_t or something. it seems like if you know it's opaque you might be more likely to look into what exactly it's supposed to be pointing to.... maybe that's just how I think about these though.

maybe we could prefix the typealiases with `Opaque`? `OpaqueSessionOptsPointer` etc.

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/ClientSession.swift]: another thought I had on these is that in cases we only use them one time, there is probably another way to get the same kind of readability without introducing an extra symbol.
for example, in this case we can just add a parameter label in the closure, like 
```swift
private func withSessionOpts<T>(wrapping options: ClientSessionOptions?,
                                _ body: (_ sessionOptsPtr: OpaquePointer) throws -> T) rethrows -> T {
// ...
}
```

[the remark this answers is not in the document]
Comment (kmahar) [reply Sources/MongoSwift/ClientSession.swift]: another thing to consider is actually just naming stored opaque pointer properties in a way that describes what they are -- for ex., `ClientSession` below has 
```swift
 internal fileprivate(set) var _session: MutableClientSessionPointer?
```

if we renamed `_session` -> `mutableClientSessionPointer` then you would get the readability benefit everyplace you use it (for ex. the `init`, in `advanceClusterTime`), not just at the place where you define the property.

then again we generally don't include the type name in other stored properties, so, idk. just a thought.

[the remark this answers is not in the document]
Comment (mbroadst) [reply Sources/MongoSwift/ClientSession.swift]: Let's just leave it as it. If the pointer is being used in difficult situations (like the ones needed for Swift 5 support) then we'll type alias them, otherwise use `OpaquePointer`

File: Sources/MongoSwift/MongoCollection+Indexes.swift (hunk 6 above)
Comment (patrickfreed) [reply Sources/MongoSwift/MongoCollection+Indexes.swift]: Updated this to use the new combine helper. It has a weird cast that needs to be there so the compiler can infer the generic type, even though it will be nil.

File: Sources/MongoSwift/ClientSession.swift (hunk 13 above)
Comment (patrickfreed) [reply Sources/MongoSwift/ClientSession.swift]: The docs you linked to deal with server sessions, which the end user of the driver doesn't necessarily need to know about and don't really inform the usage of `ClientSession`. Unfortunately, the mongodb documentation on client sessions is neither detailed nor centralized. The best summary is in the 3.6 patch notes, though they're already outdated since there are no mentions of transactions.

So as it is right now, the client session doesn't actually do anything yet. I think as we add features built on client sessions (e.g. causal consistency, transactions), we can add to the `ClientSession` documentation. I think the [python docs](http://api.mongodb.com/python/current/api/pymongo/client_session.html) have a good page on them documented in that manner.

At any rate, I've added a sentence explaining what their purpose is, which is a little better.

File: Sources/MongoSwift/ClientSession.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (patrickfreed) [reply Sources/MongoSwift/ClientSession.swift]: In line with this, I've removed the `ClientSessionMutablePointer` typealias.

File: Sources/MongoSwift/MongoCollection+Indexes.swift [hunk 15]
```diff
@@ -306,13 +327,16 @@ extension MongoCollection {
...
      * - Returns: A `MongoCursor` over the index names.
      *
-     * - Throws: A `userError.invalidArgumentError` if the options passed are an invalid combination.
+     * - Throws: `UserError.logicError` if the provided session is inactive.
      */
-    public func listIndexes() throws -> MongoCursor<Document> {
-        guard let cursor = mongoc_collection_find_indexes_with_opts(self._collection, nil) else {
+    public func listIndexes(session: ClientSession? = nil) throws -> MongoCursor<Document> {
+        // need this cast to infer a generic type
+        let opts = try encodeOptions(options: nil as DropIndexOptions?, session: session, using: self.encoder)
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCollection+Indexes.swift]: lol TIL you can cast nil... makes sense but kind of silly. did you consider adding a default value of nil for the options parameter to `encodeOptions` instead?

File: Sources/MongoSwift/MongoCursor.swift [hunk 16]
```diff
@@ -1,9 +1,13 @@
 import mongoc
 
+internal typealias MutableCursorPointer = OpaquePointer
```
Comment (kmahar) [inline Sources/MongoSwift/MongoCursor.swift]: per our other convo should maybe not be a typealias?

File: Sources/MongoSwift/Operations/Operation.swift [hunk 17]
```diff
@@ -6,3 +6,16 @@ internal protocol Operation {
     /// Executes this operation and returns its corresponding result type.
     func execute() throws -> OperationResult
 }
+
+/// Internal function for generating an options `Document` for passing to libmongoc.
+internal func encodeOptions<T: Encodable>(options: T?,
```
Comment (kmahar) [inline Sources/MongoSwift/Operations/Operation.swift]: since passing the encoder to this function is going to copy it anyway, and we don't care about the strategies when encoding options (and in the future we may introduce strategies for numbers that could mess up options encoding), maybe consider just creating the encoder here instead?

Comment (patrickfreed) [reply Sources/MongoSwift/Operations/Operation.swift]: yeah, I think that makes sense. The only reason for using the strategies here would be to encode some user data type, and if a user data type ever shows up in an options struct it'll be in the form of a `Document` anyways, meaning they could just encode it themselves. I don't really see us genericizing any options types, which would be the only way a non-`Document` user data element ends up in the options I think.

tl;dr I made it use `BSONEncoder()`
